File: lab15_naive_bayes.py
# Title     : Naive Bayes
# Objective : Conhecer os recursos que a linguagem Python tem a nos oferecer para trabalharmos com Calssificação
# Created by: accol
# Created on: 04/06/2020


import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import confusion_matrix, accuracy_score
from yellowbrick.classifier import ConfusionMatrix


# Trabalharemos com a já conhecida base de dados Cred.csv e vamos explorar o Algoritmo de Classificação Naïve Bayes
credito = pd.read_csv('D:/Users/Angelo/AULAS/Dados/credit.csv')
# Fazendo uma rápida inspeção nos dados
print('\nA dimensão da nossa base é: ', credito.shape)
print('\nOs primeiros registro da nossa base são\n', credito.head())
print('\nUm resumo da nossa base é visto a seguir\n', credito.describe())
# Primeiramente vamos definir duas varíaveis -> a primeira com os atributos previsores (20) e a segunda com o atributo classificados (1 => Class "good" or "bad"). Usaremos o método .values para já converter na nossa variável para um array numpy
previsores = credito.iloc[:, 0:20]
print('\nConferindo o tipo de nossa variável previsores: ', previsores.head())
previsores = credito.iloc[:, 0:20].values
print('\nConferindo o tipo de nossa variável: ', type(previsores))
classe = credito.iloc[:,20].values
# Observe que essa biblioteca (GaussianNB) não trabalha com dados categóricos -> requisito importante para nosso classificador, sendo assim, faremos o import de uma nova biblioteca => LabelEncoder
# Criaremos uma variável do tipo LabelEncoder
labelencoder = LabelEncoder()
# Atenção será necessário trabalhar para cada atributo categórico de sua base de dados
print('\nComo exemplo, vamos transformar o primeiro atributo categórico de nossa base')
previsores[:, 0] = labelencoder.fit_transform(previsores[:, 0])
print(previsores[:, 0])
# Vamos agora trabalhar todos os atributos categóricos de nossa base
previsores[:, 2] = labelencoder.fit_transform(previsores[:, 2])
previsores[:, 3] = labelencoder.fit_transform(previsores[:, 3])
previsores[:, 5] = labelencoder.fit_transform(previsores[:, 5])
previsores[:, 6] = labelencoder.fit_transform(previsores[:, 6])
previsores[:, 8] = labelencoder.fit_transform(previsores[:, 8])
previsores[:, 9] = labelencoder.fit_transform(previsores[:, 9])
previsores[:, 11] = labelencoder.fit_transform(previsores[:, 11])
previsores[:, 13] = labelencoder.fit_transform(previsores[:, 13])
previsores[:, 14] = labelencoder.fit_transform(previsores[:, 14])
previsores[:, 16] = labelencoder.fit_transform(previsores[:, 16])
previsores[:, 18] = labelencoder.fit_transform(previsores[:, 18])
previsores[:, 19] = labelencoder.fit_transform(previsores[:, 19])
print('\nVamos observar o primeiro registro apenas para inspecionar a transformação:\n', previsores[0])
# Nosso próximo passo será criarmos as varíaveis de treino e teste -> para na sequência trabalharmos com o Algoritmo Naïve Bayes => usaremos random_state=0 para fixar e tornar repetível os resultados desse laboratório. Obsrve que craiamos a variável X_treinamento e a variável de resposta y_treinamento e assim por diante
X_treinamento, X_teste, y_treinamento, y_teste = train_test_split(previsores, classe, test_size=0.3, random_state=0)
print('\nVamos observar nossas variáveis criadas\n')
print('\nX_treinamento\n', X_treinamento)
print('\nX_teste\n', X_teste)
# <$> Até fizemos o tratamento necessário dos dados <$>
# Vamos agora criar um objeto do tipo Naïve Bayes
naive_bayes = GaussianNB()
# Vamos agora fazer o treinamento de nosso objeto
naive_bayes.fit(X_treinamento, y_treinamento)
# A partir daqui estamos prontos para fazer as previsões -> usaremos a variável X_teste
previsoes = naive_bayes.predict(X_teste)
print('\nVamos conferir nosso previsor\n', previsoes)
print('\nA dimensão de nosso previsor é: ', previsoes.shape)
# Observe que nosso previsor usará o nosso modelo treinado para encontrar a classe "good" or "bad"
# Para avaliar a qualidade de nosso previsor devemos comparar com nossa variável y_teste que contém os dados históicos para estes 300 registros. Para visualizarmos a qualidade de nossos resultados vamo precisar de um novoimport => confusion_matrix e accuracy_score
print('\nVamos iniciar criando a matriz de confusão \n')
confusao = confusion_matrix(y_teste, previsoes)
print(confusao)
taxa_acerto = accuracy_score(y_teste, previsoes)
print('\nVamos agora calular a taxa de acerto de nosso previsor: ', round(taxa_acerto * 100, 2), '%')
taxa_erro = 1 - taxa_acerto
print('\nVamos agora calular a taxa de erro de nosso previsor:   ', round(taxa_erro * 100, 2), '%')
# Para melhorar a visão de nossa matriz de confusão precisaremos de uma nova biblioteca -> yellowbrick
# Observe que teremos que reconstruir o que foi feito anteriormente -> na prática você poderá optar por um ou outro método
v = ConfusionMatrix(GaussianNB())
v.fit(X_treinamento, y_treinamento)
v.score(X_teste, y_teste)
print('\nVamos agora fazer uma inspeção visual em nossa nova Matriz de Confusão \n')
v.poof()
# <$> Vamos agora levar nosso modelo para o ambiente de produção Simulando com registros novos => usaremos nossa base NovoCredit.csv que contém apenas um registro <$>
# Vamos precisar ler nosso arquivo e formatá-lo usando o .values
novo_credito = pd.read_csv('D:/Users/Angelo/AULAS/Dados/NovoCredit.csv')
novo_credito = novo_credito.iloc[:, 0: 20].values
# Lembre-se que deverá tratar seus dados e deixá-los todos no mesmo formato do modelo. Precisamos agora trabalhar todos os atributos categóricos de nossa base
novo_credito[:, 0] = labelencoder.fit_transform(novo_credito[:, 0])
novo_credito[:, 2] = labelencoder.fit_transform(novo_credito[:, 2])
novo_credito[:, 3] = labelencoder.fit_transform(novo_credito[:, 3])
novo_credito[:, 5] = labelencoder.fit_transform(novo_credito[:, 5])
novo_credito[:, 6] = labelencoder.fit_transform(novo_credito[:, 6])
novo_credito[:, 8] = labelencoder.fit_transform(novo_credito[:, 8])
novo_credito[:, 9] = labelencoder.fit_transform(novo_credito[:, 9])
novo_credito[:, 11] = labelencoder.fit_transform(novo_credito[:, 11])
novo_credito[:, 13] = labelencoder.fit_transform(novo_credito[:, 13])
novo_credito[:, 14] = labelencoder.fit_transform(novo_credito[:, 14])
novo_credito[:, 16] = labelencoder.fit_transform(novo_credito[:, 16])
novo_credito[:, 18] = labelencoder.fit_transform(novo_credito[:, 18])
novo_credito[:, 19] = labelencoder.fit_transform(novo_credito[:, 19])
print('\nVamos a uma rápida inspeção se tudo está ok\n', novo_credito)
# Agora vamos avaliar se esse novo cliente será um bom ou mal pagador
novo = naive_bayes.predict(novo_credito)
print('\nO novo cliente foi classificado como: ', novo)
# Assim, o algoritmo recomenda que se conceda o crédito!

File: lab15_naive_bayes.R
# Title     : Naive Bayes
# Objective : Conhecer os recursos que a linguagem R tem a nos oferecer para trabalharmos com Calssificação
# Created by: accol
# Created on: 04/06/2020

# Importante -> vamos precisar instalar dois pacotes que juntos respondem por praticamente tudo que R oferece para trabalhar com Machine Learning -> são eles: e1071 e klaR

# Para este laboratório usaremos o pacote e1071 e a base de dados Credit.csv -> o separador default é a ',', logo desnecessário, será aqui empregado para que se lembre em caso de um separador diferente deverá ser utilizado, ok

library(e1071)

credito = read.csv('D:/Users/Angelo/AULAS/Dados/Credit.csv',',', header = T)
# Vamos iniciar fazendo uma rápida inspeção em nossa base de dados
cat('\nA base de dados Credit.csv é composta por\n')
credito
# Observe que o último atributo é a Classe que nos diz se ele/ela é um bom ou mal pagador
cat('\nObservando os primeiros registros de nossa base:\n ')
head(credito)
cat('\nQual a dimensão desta base\n', dim(credito))
cat('\nUm sumário da base para início dos trabalhos\n')
summary(credito)
# Para este laboratório queremos prever se um dado cliente solicitando crédito será um bom ou mal pagador => para isso teremos que gerar um modelo a partir dos dados históricos >>= usaremos 70% dos dados para gerar nosso modelo e 30% para realizar os testes do modelo ||> usaremos para esse propóstio a função sample() -> estudada anteriormente
# Para paramtrizar nossa função sample() usaremos -> parâmetro 1 = 2 (siginfica que estaremos gerando 1 ou 2, parâmetro 2 = 1000 (temos 1000 registros em nossa base), parâmetro 3 = replace = T (será com reposição), parâmetro 4 = prob=c(0.7, 0.3) ||> trabalharemos com 70% e 30% -> acompanhe
amostra = sample(2, 1000, replace = T, prob = c(0.7, 0.3))
cat('\nVisualizando nossa amostra que deverá ser um array composto por 1000 registros de 1 (70%) e 2 (30%)')
amostra
# Precisamos agora separar nossa amostra em dois conjuntos de dados um para treino e outro para teste -> acompanhe
creditotreino = credito[amostra == 1, ]
creditoteste = credito[amostra == 2, ]
cat('\nQual a dimensão da variável creditotreino\n', dim(creditotreino))
cat('\nInspecionando nossa variável creditotreino\n')
creditotreino
cat('\nQual a dimensão da variável teste\n', dim(creditoteste))
cat('\nInspecionando nossa variável creditoteste\n')
creditoteste
# Observe que as variáveiis creditotreino e creditoteste não apresentam resultados exatos 70% e 30% respectivamente, isso porque nossa função usa probabilidades independentes -> caso de dúvidas revise o lab onde se trabalhou com a função sample()
# O importante é que a separação dos dados em treino e teste foram realizados de forma aleatória ||> Agora vamos gerar nosso modelo -> usaremos o algoritmo naiveBayes
# Observe que estaremos usando a class (good or bad -> será nossa variável de resposta) o ~ (separa os atributos -> mostrando a separação entre as variáveis explicativas e a variável de resposta) e o . (que representa todos os argumentos -> serão nossas variáveis explicativas) ||> se quisesse uma seleção de atributos (variáveis explicativas) teria que inserir um a um concatenando-os com o sinal de '+'
modelo = naiveBayes(class ~ ., creditotreino)
cat('\nQual a dimensão do nosso modelo\n', dim(modelo))
cat('\nVamos fazer uma ispeção em nosso modelo\n')
cat('\nPodemos conferir a classe do nosso modelo: ', class(modelo))
modelo
# Atenção -> outro algoritmo poderá gerar outro resultado para seu modelo
# Precsiamos agora criar nossa variável teste para avaliar nosso modelo ||> note que a função para realizar previsões será sempre a mesma >>= predict -> que precisa de dois parâmentros um o modelo e o outro a variável de teste
predicao = predict(modelo, creditoteste)
cat('\nPodemos conferir nosso array predição\n ')
predicao
# Feito isso teremos um vetor predicao com aproximadamente 30% dos dados e podemos agora comparar esse resultado com o a amostra original que contém os dados históricos conhecidos
# Para essa análise vamos criar uma matriz de confusão -> acompanhe
confusao = table(creditoteste$class, predicao)
cat('\nPodemos conferir o resultado analisando a marriz de confusão\n ')
confusao
# Para melhor avaliar estes resultados vamos calcular o índice de acertos => acompanhe pegaremos a célula 1 e célula 4 da matriz de confusão que corresponde aos acertos e dividiremos pelo total de registros, observe
taxaacerto = (confusao[1] + confusao[4]) / sum(confusao)
cat('\nA taxa de acerto do nosso modelo é de: ', round(taxaacerto * 100, 2),'%')
# Se quisermos calcular a taxa de erro basta:
taxaerro = (confusao[2] + confusao[3]) / sum(confusao)
cat('\nA taxa de acerto do nosso modelo é de: ', round(taxaerro * 100, 2),'%')
# A partir desse momento cabe o analista avaliar se o modelo está bom o suficiente para ser colocado no ambiente de produção. Poderá o analista trabalhar o modelo com base nos fundamentos apresentados a fim de melhorar sua resposta a um ponto que considere aceitável para seu projeto
# Vamos neste momento simular o Modelo em Produção
# Para nosso teste em Produção precisamos de um/vários novo cliente/s para submetê-lo ao modelo e verificar sua resposta
novocredito = read.csv('D:/Users/Angelo/AULAS/Dados/NovoCredit.csv', sep = ',', header = T)
# Vamos iniciar fazendo uma rápida inspeção em nossa base de dados -> NovoCredit
cat('\nA base de dados NovoCredit.csv é composta por\n')
novocredito
cat('\nQual a dimensão da base NovoCredt\n', dim(novocredito))
# Observe que esta base não apresenta o classe -> temos os 20 primeiros atributos tal e qual o modelo, mas a Classe não esta presente (ela nosso modelo deverá predizer)
cat('\nObservando os primeiros registros de nossa novocredito:\n ')
head(novocredito)
# Vamos agora trabalhar a previsão dos dados
predicao_novo = predict(modelo, novocredito)
cat('\nPodemos conferir nosso array predição_novo\n ')
predicao_novo
